gtsam/hybrid/tests/testGaussianMixtureFactor.cpp

```diff
@@ -191,8 +196,240 @@ TEST(GaussianMixtureFactor, Error) {
+// Test components with differing means
+TEST(GaussianMixtureFactor, DifferentMeans) {
+  DiscreteKey m1(M(1), 2), m2(M(2), 2);
+
+  Values values;
+  double x1 = 0.0, x2 = 1.75, x3 = 2.60;
+  values.insert(X(1), x1);
+  values.insert(X(2), x2);
+  values.insert(X(3), x3);
+
+  auto model0 = noiseModel::Isotropic::Sigma(1, 1e-0);
+  auto model1 = noiseModel::Isotropic::Sigma(1, 1e-0);
+  auto prior_noise = noiseModel::Isotropic::Sigma(1, 1e-0);
+
+  auto f0 = std::make_shared<BetweenFactor<double>>(X(1), X(2), 0.0, model0)
+                ->linearize(values);
+  auto f1 = std::make_shared<BetweenFactor<double>>(X(1), X(2), 2.0, model1)
+                ->linearize(values);
+  std::vector<GaussianFactor::shared_ptr> factors{f0, f1};
+
+  GaussianMixtureFactor mixtureFactor({X(1), X(2)}, {m1}, factors, true);
+  HybridGaussianFactorGraph hfg;
+  hfg.push_back(mixtureFactor);
+
+  f0 = std::make_shared<BetweenFactor<double>>(X(2), X(3), 0.0, model0)
+           ->linearize(values);
+  f1 = std::make_shared<BetweenFactor<double>>(X(2), X(3), 2.0, model1)
+           ->linearize(values);
+  std::vector<GaussianFactor::shared_ptr> factors23{f0, f1};
+  hfg.push_back(GaussianMixtureFactor({X(2), X(3)}, {m2}, factors23, true));
+
+  auto prior = PriorFactor<double>(X(1), x1, prior_noise).linearize(values);
+  hfg.push_back(prior);
+
+  hfg.push_back(PriorFactor<double>(X(2), 2.0, prior_noise).linearize(values));
+
+  auto bn = hfg.eliminateSequential();
+  HybridValues actual = bn->optimize();
+
+  HybridValues expected(
+      VectorValues{
+          {X(1), Vector1(0.0)}, {X(2), Vector1(0.25)}, {X(3), Vector1(-0.6)}},
+      DiscreteValues{{M(1), 1}, {M(2), 0}});
+
+  EXPECT(assert_equal(expected, actual));
+
+  {
+    DiscreteValues dv{{M(1), 0}, {M(2), 0}};
+    VectorValues cont = bn->optimize(dv);
+    double error = bn->error(HybridValues(cont, dv));
+    // regression
+    EXPECT_DOUBLES_EQUAL(1.77418393408, error, 1e-9);
+  }
+  {
+    DiscreteValues dv{{M(1), 0}, {M(2), 1}};
+    VectorValues cont = bn->optimize(dv);
+    double error = bn->error(HybridValues(cont, dv));
+    // regression
+    EXPECT_DOUBLES_EQUAL(1.77418393408, error, 1e-9);
+  }
+  {
+    DiscreteValues dv{{M(1), 1}, {M(2), 0}};
+    VectorValues cont = bn->optimize(dv);
+    double error = bn->error(HybridValues(cont, dv));
+    // regression
+    EXPECT_DOUBLES_EQUAL(1.10751726741, error, 1e-9);
+  }
+  {
+    DiscreteValues dv{{M(1), 1}, {M(2), 1}};
+    VectorValues cont = bn->optimize(dv);
+    double error = bn->error(HybridValues(cont, dv));
+    // regression
+    EXPECT_DOUBLES_EQUAL(1.10751726741, error, 1e-9);
+  }
+}
+
+/* ************************************************************************* */
+/**
+ * @brief Test components with differing covariances.
+ * The factor graph is
+ *     *-X1-*-X2
+ *          |
+ *          M1
+ */
+TEST(GaussianMixtureFactor, DifferentCovariances) {
```

Comment:
@dellaert see this test as an example vs DifferentCovariances2. Specifying the FG is as easy as in continuous case.

Reply:
Same comment. I don’t think this is easy. The HBN -> HFG is easy.